### ISSUE TITLE
QY MiniCPM5-2B: --enable-graph decode loops on FIM special tokens (LongBench CoT EM=0)

### ISSUE THREAD

**Issue by Ceng23333** [Ceng23333/InfiniTensorWorktree#3]
## Summary

On Denglin QY, InfiniLM `--enable-graph` decode for MiniCPM5-2B is corrupted: LongBench CoT (`MAX_INPUT=7800`, serve `max-cache-len=16384`) scored **EM=0.0000** with **59/59 `pred=None`**. Completions start with `The` then loop the MiniCPM FIM special token `<|fim_prefix|>` until `completion_tokens≈1150`.

The same model/data/host are fine without InfiniLM graph capture, and vLLM with CUDA graphs FULL on the same host scored **EM=0.3729**. This is an InfiniLM(+InfiniCore) **graph-path** bug, not a model or dataset issue.

**Keep this issue open** until a new InfiniTensorWorktree pin is cut and verified. Link later InfiniLM / InfiniCore commits and PRs here.

## Pin

| Component | Branch / tag | SHA |
|---|---|---|
| InfiniTensorWorktree | `release/v2026.09.15-qy-minicpm5` / tag `v2026.09.15-qy-minicpm5` | `ada3d5721f728ee462e353d373023b10008e4cd4` |
| InfiniLM (live, used in repro) | `fix/qy-minicpm5-v2026.09.15` | `8343f5ed5b028937bfdbdc7f46be1c959be8016c` (includes `chat_template_kwargs` forward) |
| InfiniCore | `fix/qy-minicpm5-v2026.09.15` | `9b67084ce3a21cb3326404f768198f026852a23c` |
| Checkout | `worktrees/InfiniTensorWorktree-v029` | do not mutate the published `v2026.09.15-qy-minicpm5` tag |

- Device: **QY** (`--device qy`, Denglin `dlrt`)
- `max-cache-len=16384`, `max-new-tokens=2048`, `tp=1`, `temperature=0`, `top-p=1`, `top-k=1`
- Model: MiniCPM5-2B (`OpenBMB/MiniCPM5-2B`)
- Chat: LongBench official CoT, `chat_template_kwargs={'enable_thinking': False}`

## Reproduction

Launch InfiniLM **with** `--enable-graph` (broken) vs omit the flag (control):

```bash
# via playground wrap (default ENABLE_GRAPH is currently 0 after this regression)
SUPPORT_ROOT=/home/qinyiqun/workspace/minicpm5-2b-support
IO_PG=/home/qinyiqun/workspace/InfiniOrchestrator/playground/Standalone
export MAX_CACHE_LEN=16384 DENGLIN_DEVICES=2

ENABLE_GRAPH=1 "${IO_PG}/minicpm5-2b-qy-inf/run-wrap.sh"   # broken
# ENABLE_GRAPH=0 "${IO_PG}/minicpm5-2b-qy-inf/run-wrap.sh" # control
```

Equivalent server argv:

```bash
python3 /workspace/InfiniLM/python/infinilm/server/inference_server.py \
  --device qy \
  --model=/models/minicpm5-2b \
  --tp=1 --host=0.0.0.0 --port=18001 \
  --max-new-tokens=2048 --max-batch-size=1 --max-cache-len=16384 \
  --temperature=0 --top-p=1 --top-k=1 \
  --enable-graph
```

Then run LongBench CoT `short`/`easy` (`n=59`, `MAX_INPUT=7800`, official CoT). A shorter probe also suffices: same chat template, `enable_thinking=false`, ~2–4k prompt, `max_tokens=64`, compare graph on vs off.

## Symptom

Every request in the graph-on suite looks like:

- `_id=66ec1de5821e116aacb1adca`, gold `answer=D`, **`pred=None`**, `truncated=True`
- `completion_tokens=1150` (all 59 rows)
- `prompt_tokens` ~8931–9344 (mean 9034.7)
- `response` starts `The` then repeats `<|fim_prefix|>` (sample: `The<|fim_prefix|><|fim_prefix|>…`)

No A–D choice is ever produced, so the judge records `None`.

## Artifacts

| Leg | Suite | Result |
|---|---|---|
| **Inf + `--enable-graph` (broken)** | `InfiniOrchestrator/bench_results/bench_client_minicpm5-2b_inf_cot_base8k_20260920_135348` | **EM=0.0000**, 59/59 `pred=None`, all `completion_tokens=1150`, FIM loops |
| **vLLM CUDA graphs FULL (same host, same CoT / MAX_INPUT=7800)** | `bench_results/bench_client_minicpm5-2b_vllm_cot_base8k_20260920_135348` | **EM=0.3729**, none=0 |
| **Inf `enable_graph=False` (prior CoT, MAX_INPUT=6000)** | `bench_results/bench_client_minicpm5-2b_inf_cot_20260918_210216` | **EM=0.4237** |
| **Inf `enable_graph=False` (same base8k CoT rerun)** | `bench_results/bench_client_minicpm5-2b_inf_cot_base8k_nograph_20260920_154237` | control rerun (in flight / baseline) |

Broken preds: `…/longbench_v2/longbench_preds.jsonl`. Summary: `…/longbench_v2/longbench_summary.json` (`lb_em=0.0`, `lb_n=59`, `lb_truncated_n=59`).

Playground wrap that currently defaults graph **off** because of this: `InfiniOrchestrator/playground/Standalone/minicpm5-2b-qy-inf/run-wrap.sh`. Restore `ENABLE_GRAPH=1` only after acceptance below.

## Expected / acceptance

Graph semantics must be **correct**, not silently disabled:

1. **Short probe** (CoT and noCoT): `--enable-graph` yields normal A–D answers; **no** `<|fim_prefix|>` (or other FIM special-token) loops; graph-on text matches graph-off on the same prompt.
2. **Regression check**: relaunch Inf wrap with `ENABLE_GRAPH=1`; rerun Inf CoT only `short_easy` / `MAX_INPUT=7800`; expect EM near prior no-graph Inf CoT (~0.42) and **zero** `None` preds from FIM garbage.
3. Land IL/(IC) SHAs, then a new ITW pin (planned tag `v2026.09.20-qy-minicpm5-cudagraph`). Close this issue only after that pin is verified with CoT EM + SHAs.

## Suspected area (for the follow-up fix)

Stop at the first confirmed cause:

- Sampling / vocab: graph decode ignores logits processors, stop tokens, or request temperature/top_k
- Special-token masking: MiniCPM FIM tokens emitted as normal tokens under graph replay
- Static KV + graph: capture shapes vs variable prompt length on QY (`InferEngine` → InfiniCore `src/infinicore/graph/graph.cc`)
- Prefill vs decode graph split: corrupt state when replaying decode after long prefill
